app/lib/search/backend.dart

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+
+import '../frontend/models.dart';
+import '../shared/mock_scores.dart';
+import '../shared/search_service.dart';
+
+/// Sets the backend service.
+void registerSearchBackend(SearchBackend backend) =>
+    ss.register(#_searchBackend, backend);
+
+/// The active backend service.
+SearchBackend get searchBackend => ss.lookup(#_searchBackend);
+
+/// Datastore-related access methods for the search service
+class SearchBackend {
+  final DatastoreDB _db;
+
+  SearchBackend(this._db);
+
+  Future<List<PackageDocument>> loadDocuments(List<String> packageNames) async {
+    final List<Key> packageKeys = packageNames
+        .map((String name) => _db.emptyKey.append(Package, id: name))
+        .toList();
+    final List<Package> packages = await _db.lookup(packageKeys);
+    packages.removeWhere((p) => p == null);
+    if (packages.isEmpty) return [];
+
+    final List<Key> versionKeys =
+        packages.map((p) => p.latestVersionKey).toList();
+    final List<PackageVersion> versions = await _db.lookup(versionKeys);
+
+    final List<PackageDocument> documents = [];
+    for (int i = 0; i < packages.length; i++) {
+      final Package p = packages[i];
+      final PackageVersion pv = versions[i];
+      if (pv == null) continue;
+
+      documents.add(new PackageDocument(
+        url: _toUrl(pv.package),
+        package: pv.package,
+        version: pv.version,
+        devVersion: p.latestDevVersion,
+        detectedTypes: pv.detectedTypes,
+        description: pv.pubspec.description,
+        lastUpdated: pv.shortCreated,
+        readme: pv.readmeContent,
+        popularity: mockScores[pv.package] ?? 0.0,
+      ));
+    }
+    return documents;
+  }
+
+  Stream<String> listPackages({DateTime updatedAfter}) {
+    return _db
+        .query(Package)
```

Comment:
You should be able to use filters:
```dart
Stream<String> listPackages({DateTime updatedAfter}) {
  final query = db.query(Package);
  if (updatedAfter != null) query.filter('>= updated', updatedAfter);
  for (final Package package in query.run()) {
    yield package.name;
  }
}
```

So the datastore can do the filtering instead of querying everything and doing filtering in Dart.